### ISSUE TITLE
Calling invokeAndWait from read-action leads to possible deadlock.

### ISSUE THREAD

**Issue by joshuataylor** [KronicDeth/intellij-elixir#959]
# Version

7.1.0

# Exception

## Message

Calling invokeAndWait from read-action leads to possible deadlock.

## Stacktrace

```
java.lang.IllegalStateException: Calling invokeAndWait from read-action leads to possible deadlock.
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:658)
	at org.elixir_lang.sdk.elixir.Type.putDefaultErlangSdk(Type.java:381)
	at org.elixir_lang.mix.runner.MixRunningStateUtil.setElixir(MixRunningStateUtil.java:332)
	at org.elixir_lang.mix.runner.MixRunningStateUtil.commandLine(MixRunningStateUtil.java:204)
	at org.elixir_lang.mix.runner.MixRunningStateUtil.commandLine(MixRunningStateUtil.java:196)
	at org.elixir_lang.credo.Annotator.generalCommandLine(Annotator.java:126)
	at org.elixir_lang.credo.Annotator.generalCommandLine(Annotator.java:169)
	at org.elixir_lang.credo.Annotator.generalCommandLine(Annotator.java:320)
	at org.elixir_lang.credo.Annotator.doAnnotate(Annotator.java:291)
	at org.elixir_lang.credo.Annotator.doAnnotate(Annotator.java:45)
	at com.intellij.codeInspection.ExternalAnnotatorInspectionVisitor.checkFileWithExternalAnnotator(ExternalAnnotatorInspectionVisitor.java:70)
	at org.elixir_lang.credo.inspection_tool.Local.checkFile(Local.kt:26)
	at com.intellij.codeInspection.LocalInspectionTool$1.visitFile(LocalInspectionTool.java:142)
	at com.intellij.extapi.psi.PsiFileBase.accept(PsiFileBase.java:70)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:82)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:309)
	at com.intellij.concurrency.ApplierCompleter.d(ApplierCompleter.java:133)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1122)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:105)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:543)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:488)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.concurrency.ApplierCompleter.a(ApplierCompleter.java:116)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:96)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:143)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:229)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:96)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.pollAndExecCC(ForkJoinPool.java:1190)
	at java.util.concurrent.ForkJoinPool.helpComplete(ForkJoinPool.java:1879)
	at java.util.concurrent.ForkJoinPool.awaitJoin(ForkJoinPool.java:2045)
	at java.util.concurrent.ForkJoinTask.doJoin(ForkJoinTask.java:390)
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:719)
	at java.util.concurrent.ForkJoinPool.invoke(ForkJoinPool.java:2616)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgress(JobLauncherImpl.java:65)
	at com.intellij.concurrency.JobLauncher.invokeConcurrentlyUnderProgress(JobLauncher.java:57)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:320)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:230)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.collectInformationWithProgress(LocalInspectionsPass.java:136)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doCollectInformation(ProgressableTextEditorHighlightingPass.java:83)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:70)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.a(PassExecutorService.java:437)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1127)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.c(PassExecutorService.java:430)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:543)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:488)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.d(PassExecutorService.java:429)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.b(PassExecutorService.java:405)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:143)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:229)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:403)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:170)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

```

**Comment by KronicDeth** [comment]
If this is related to #960,  I'd recommend manually recreating your SDK instead of having the IDE upgrade it in the background as a work-around.

1. File > Project Structure
2. Platform Settings > SDKs
3. Delete any Elixir SDKs by selecting them and clicking the - button on the top of the center column
4. Delete any Erlang SDKs by selecting them and clicking the - button on the top of the center column
5. Delete any Erlang SDKs for Elixir SDKs by selecting them and clicking the - button on the top of the center column
6. Click + 
7. Select Elixir SDK from the Add New SDK menu that came from the + button
8.  Click OK in the "Cannot Create SDK" dialog that say you need to create an Erlang SDK first
9. Click Open if the IDE selected the right Erlang SDK Home Path; otherwise, navigate to the correct Home Path first
10. Click Open if the IDE selected the right SDK Home Path; otherwise, navigate to the correct Home Path first.
  When you're done, you should have an Erlang SDK (if `intellij-erlang` is installed too) or an Erlang SDK for Elixir SDK and and Elixir SDK.  The Elixir SDK should have "Internal Erlang SDK" combobox with the Erlang SDK you created selected.
11. Click Apply to save all the SDK changes; Click OK to save and close the Project Structure Dialog.

**Comment by KronicDeth** [comment]
Did you not use the `credo` functionality in 7.0.0?  That should have also been trying to update your SDK.

**Comment by joshuataylor** [comment]
No, I did not use `credo` functionality yet for 7.0.0, something I've been meaning to try. Will try above and let you know.

**Comment by KronicDeth** [comment]
I didn't mean to imply that `intellij-erlang` was necessary.  The only change is that you'll get the benefit of the Erlang plugin when looking at Erlang-only dependencies.  For the purposes of running mix, either the Erlang SDK from `intellij-erlang` or the Erlang SDK for Elixir SDK, which is a more limited one I added to `intellij-elixir` to allow the `ebin` paths from Erlang to be tracked, work the same.

Did the work-around fix #960 too?

**Comment by KronicDeth** [comment]
Hmm, from the code it looks like the message is a just a warning.

```java
if (holdsReadLock()) {
      LOG.error("Calling invokeAndWait from read-action leads to possible deadlock.");
}    
```
-- https://github.com/JetBrains/intellij-community/blob/b359c9253ac18e1d87295bec1652b683ea0ac6dd/platform/platform-impl/src/com/intellij/openapi/application/impl/ApplicationImpl.java#L658-L660

 It probably won't explain #960 unless running `credo` in your code is so-slow that it's taxing the memory or CPU allocated to IntelliJ, which some users had reported.  It's why `credo` can be dialed back in 7.0.0.
